sentry-ruby/lib/sentry/session_flusher.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Sentry
+  class SessionFlusher
+    include LoggingHelper
+
+    FLUSH_INTERVAL = 60
+
+    def initialize(configuration, client)
+      @thread = nil
+      @client = client
+      @pending_aggregates = {}
+      @release = configuration.release
+      @environment = configuration.environment
+      @logger = configuration.logger
+
+      log_debug("[Sessions] Sessions won't be captured without a valid release") unless @release
+    end
+
+    def flush
+      return if @pending_aggregates.empty?
+      envelope = pending_envelope
+
+      Sentry.background_worker.perform do
+        @client.transport.send_envelope(envelope)
+      end
+
+      @pending_aggregates = {}
+    end
+
+    def add_session(session)
+      return unless @release
+
+      ensure_thread
+
+      return unless Session::AGGREGATE_STATUSES.include?(session.status)
+      @pending_aggregates[session.started_bucket] ||= init_aggregates
+      @pending_aggregates[session.started_bucket][session.status] += 1
+    end
+
+    def kill
+      @thread&.kill
+    end
+
+    private
+
+    def init_aggregates
+      Session::AGGREGATE_STATUSES.map { |k| [k, 0] }.to_h
+    end
+
+    def pending_envelope
+      envelope = Envelope.new
+
+      header = { type: 'sessions' }
+      payload = { attrs: attrs, aggregates: aggregates_payload }
+
+      envelope.add_item(header, payload)
+      envelope
+    end
+
+    def attrs
+      { release: @release, environment: @environment }
+    end
+
+    def aggregates_payload
+      @pending_aggregates.map do |started, aggregates|
+        aggregates[:started] = started.iso8601
+        aggregates
+      end
+    end
+
+    def ensure_thread
+      return if @thread&.alive?
+
+      @thread = Thread.new do
```

Comment:
yes I should have posted this in the PR description.
Basically I looked at using `TimerTask` because what we need in this thread is a 60 second scheduler.
I found [several issues](https://github.com/ruby-concurrency/concurrent-ruby/issues?q=TimerTask) and most importantly [this one regarding thread leaking]. But if we go with just a single `ThreadPoolWorker`, I think that should be fine and I can make those changes.

Reply:
fwiw I did test the core thread logic with [this toy flusher](https://gist.github.com/sl0thentr0py/fb9558715cdebb642090ad638273fc41).

Reply:
I still think `ThreadPoolExecutor` can also be used here? Something like:

```rb
executor.post do
  loop do
    sleep Flusher::FLUSH_INTERVAL
    flusher.flush
  end
end
```

The benefits are:

- We can terminate it the same way [we do for background worker](https://github.com/getsentry/sentry-ruby/blob/cf1e5b24788fb50b3ea49ecf9b1100162e1741e6/sentry-ruby/lib/sentry/background_worker.rb#L41-L44), which is more flexible then `kill`.
- Having a unified abstraction for all the background worker/thread is better for maintainability.

And the only cons is probably that we need to make some adjustments to the `BackgroundWorker` first to make some logics sharable. But I think that's fine and I can work on it too.

Reply:
I don't know, I don't think `ThreadPoolExecutor` is meant for long-lived threads, it has 
```
    #   @option opts [Integer] :idletime (DEFAULT_THREAD_IDLETIMEOUT) the maximum
    #     number of seconds a thread may be idle before being reclaimed
```

We need a _scheduler_ thread that lives long and schedules the sending every 60s. The right abstraction for that is `TimerTask` which is unfortunately broken. `concurrent-ruby` has a lot of general abstractions with much more logic in each of them and I actually prefer this very simple implementation rather than shoe-horning logic into something that might not work.

Reply:
we have the same simple logic in the [python SDK](https://github.com/getsentry/sentry-python/blob/b63b5ae32d254771cbd027009d0b6def7e2e07fb/sentry_sdk/sessions.py#L106-L126).

Reply:
ok, let's give it a try. and just to confirm, it doesn't hang when the program exists?

Reply:
cool, if you're fine with it, I'll do some extensive testing of the branch again on Monday for threading edge cases.

Reply:
@st0012 quick note on the testing non-determinism, the way the python tests solve this is to have an explicit `flush` on the worker which gets rid of the timing problem by just waiting for everything to finish. Caveat there is that we can not test the threading/concurrent logic itself which is fairly tricky in general anyway.

Reply:
@st0012 I've added my manual testing scenarios and proofs/logs below in separate comments. Let me know if you have any questions!